cctp_primitives/src/proving_system/mod.rs

```diff
@@ -441,4 +441,165 @@ pub fn check_matching_proving_system_type(
+    let h_1_segs = ((2 * h + 4 * zk_bound - 2) as f64/segment_size as f64).ceil() as usize;
+    let z_2_segs = (k as f64/segment_size as f64).ceil() as usize;
+    let h_2_segs =  ((3 * k - 3) as f64/segment_size as f64).ceil() as usize;
+
+    let num_segments = 3 * w_z_a_b_segs + t_segs + z_1_segs + h_1_segs + h_2_segs + z_2_segs;
+
+    // Compute sizes
+    let num_evaluations = 22; // indexer polys (12) + prover polys (8) + 2 (z_1 and z_2 are queried at 2 different points) 
+
+    let pc_proof_size = 1 // l_vec_len
+        + 2 * algebra::log2_floor(segment_size) * 33 // l_vec and r_vec elems
+        + 33 // G_final
+        + 32 // c_final
+        + 1 // Hiding comm is Some or None
+        + if zk { 33 } else { 0 } // If zk we will have the hiding comm
+        + 1 // Rand is Some or None
+        + if zk { 32 } else { 0 }; // If zk we will have the rand
+
+    let batch_poly_segs = ((3 * k - 4) as f64/segment_size as f64).ceil() as usize;
+    let pc_batch_proof_size = (num_evaluations - 2) * 32 // 32 bytes to serialize 1 field element
+        + 1 // 1 byte to encode length of evaluations vec
+        + 33 * batch_poly_segs // num segs of the highest degree polynomial as the batch poly will have this degree too
+        + 1 // 1 byte to encode length of segments vec
+        + pc_proof_size as usize;
+
+    let proof_size = num_segments * 33 // 33 bytes used for point compressed representation
+        + 8 // 1 byte for each poly to encode shifted comm being Some or None
+        + 8 // 1 byte for each poly to encode length of segments vector
+        + num_evaluations * 32
+        + pc_batch_proof_size
+        + match proof_type {
+            ProvingSystem::Darlin => 
+                2 * // 2 deferred accumulators
+                (
+                    33 // G_final
+                    + 1 // xi_s len
+                    + algebra::log2_floor(segment_size) * 16 // xi_s (only 128 bits long)
+                ),
+            ProvingSystem::CoboundaryMarlin => 0,
+            _ => unreachable!()
+        } as usize;
+
+    let indexer_polys_num_segs = (k as f64/segment_size as f64).ceil() as usize;
+    let vk_size = 24 // index_info
+        + 1 // indexer comms vec len
+        + indexer_polys_num_segs * 33 * 12 // segment commitments for each indexer poly
+        + 12 // comms vec len for each indexer poly
+        + 12 // shifted comm some or none for each indexer poly
+    ;
+    
+    (proof_size, vk_size)
+}
+
+#[allow(dead_code)]
+/// Given segment_size, density, zk, proof_type, return the log2 of the maximum value of |H|
+/// s.t. proof size is <= max_proof_size and vk size is <= max_vk_size, and the
+/// corresponding values of proof size and vk size
+pub(crate) fn compute_max_domain_h_size(
+    segment_size: usize,
+    density: usize,
+    zk: bool,
+    max_proof_size: usize,
+    max_vk_size: usize,
+    proof_type: ProvingSystem
+) -> (usize, usize, usize)
```

Comment:
I'd still opt for a ``compute_max_constraints_and_variables()`` instead, as it is might be more informative in practice:
Given the density, a circuit might have more constraints than the one returned by the function. The reason for this is that the function applies the density to a power of two, which is only realistic if the circuit does in fact use all "space" given by a two-adic domain. In practice, the constraint system is padded, but this padding does not affect the non-zero entries hence the domain K needed.

Reply:
Done, please review it

Reply:
```
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 131072, Max supported variables are: 131072, Proof size: 5389 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 262144, Max supported variables are: 262144, Proof size: 5455 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 524288, Max supported variables are: 524288, Proof size: 5521 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 1048576, Max supported variables are: 1048576, Proof size: 5587 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 131072, Max supported variables are: 131072, Proof size: 5192 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 262144, Max supported variables are: 262144, Proof size: 5258 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 524288, Max supported variables are: 524288, Proof size: 5324 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 1048576, Max supported variables are: 1048576, Proof size: 5390 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 131072, Max supported variables are: 131072, Proof size: 5937 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 262144, Max supported variables are: 262144, Proof size: 6035 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 524288, Max supported variables are: 524288, Proof size: 6133 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 1048576, Max supported variables are: 1048576, Proof size: 6231 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 131072, Max supported variables are: 131072, Proof size: 5740 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 262144, Max supported variables are: 262144, Proof size: 5838 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 524288, Max supported variables are: 524288, Proof size: 5936 bytes, Vk size: 3225 bytes
 
For Density: 2, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 1048576, Max supported variables are: 1048576, Proof size: 6034 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 87381, Max supported variables are: 131072, Proof size: 5389 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 174762, Max supported variables are: 262144, Proof size: 5455 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 349525, Max supported variables are: 524288, Proof size: 5521 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 699050, Max supported variables are: 1048576, Proof size: 5587 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 87381, Max supported variables are: 131072, Proof size: 5192 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 174762, Max supported variables are: 262144, Proof size: 5258 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 349525, Max supported variables are: 524288, Proof size: 5324 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 699050, Max supported variables are: 1048576, Proof size: 5390 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 87381, Max supported variables are: 131072, Proof size: 5937 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 174762, Max supported variables are: 262144, Proof size: 6035 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 349525, Max supported variables are: 524288, Proof size: 6133 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 699050, Max supported variables are: 1048576, Proof size: 6231 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 87381, Max supported variables are: 131072, Proof size: 5740 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 174762, Max supported variables are: 262144, Proof size: 5838 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 349525, Max supported variables are: 524288, Proof size: 5936 bytes, Vk size: 3225 bytes
 
For Density: 3, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 699050, Max supported variables are: 1048576, Proof size: 6034 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 65536, Max supported variables are: 65536, Proof size: 4927 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 131072, Max supported variables are: 131072, Proof size: 4993 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 262144, Max supported variables are: 262144, Proof size: 5059 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 524288, Max supported variables are: 524288, Proof size: 5125 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 65536, Max supported variables are: 65536, Proof size: 4730 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 131072, Max supported variables are: 131072, Proof size: 4796 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 262144, Max supported variables are: 262144, Proof size: 4862 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 524288, Max supported variables are: 524288, Proof size: 4928 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 65536, Max supported variables are: 65536, Proof size: 5475 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 131072, Max supported variables are: 131072, Proof size: 5573 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 262144, Max supported variables are: 262144, Proof size: 5671 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 524288, Max supported variables are: 524288, Proof size: 5769 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 65536, Max supported variables are: 65536, Proof size: 5278 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 131072, Max supported variables are: 131072, Proof size: 5376 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 262144, Max supported variables are: 262144, Proof size: 5474 bytes, Vk size: 3225 bytes
 
For Density: 4, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 524288, Max supported variables are: 524288, Proof size: 5572 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 52428, Max supported variables are: 65536, Proof size: 4927 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 104857, Max supported variables are: 131072, Proof size: 4993 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 209715, Max supported variables are: 262144, Proof size: 5059 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: true, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 419430, Max supported variables are: 524288, Proof size: 5125 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 52428, Max supported variables are: 65536, Proof size: 4730 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 104857, Max supported variables are: 131072, Proof size: 4796 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 209715, Max supported variables are: 262144, Proof size: 4862 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: CoboundaryMarlin, Zk: false, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 419430, Max supported variables are: 524288, Proof size: 4928 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 52428, Max supported variables are: 65536, Proof size: 5475 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 104857, Max supported variables are: 131072, Proof size: 5573 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 209715, Max supported variables are: 262144, Proof size: 5671 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: true, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 419430, Max supported variables are: 524288, Proof size: 5769 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 15, Num inputs: 32, Max supported constraints are: 52428, Max supported variables are: 65536, Proof size: 5278 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 16, Num inputs: 32, Max supported constraints are: 104857, Max supported variables are: 131072, Proof size: 5376 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 17, Num inputs: 32, Max supported constraints are: 209715, Max supported variables are: 262144, Proof size: 5474 bytes, Vk size: 3225 bytes
 
For Density: 5, MaxProofSize: 7000, MaxVkSize: 4000, ProofType: Darlin, Zk: false, SegmentSize: 1 << 18, Num inputs: 32, Max supported constraints are: 419430, Max supported variables are: 524288, Proof size: 5572 bytes, Vk size: 3225 bytes
```